sdks/python/apache_beam/transforms/util.py

```diff
@@ -864,3 +867,225 @@ def add_window_info(element, timestamp=DoFn.TimestampParam,
 
     def expand(self, pcoll):
       return pcoll | ParDo(self.add_window_info)
+
+
+class Regex(object):
+  """
+  PTransform  to use Regular Expression to process the elements in a
+  PCollection.
+  """
+  @staticmethod
+  def _regex_compile(regex):
+    """Return re.compile if the regex has a string value"""
+    if isinstance(regex, str):
+      regex = re.compile(regex)
+    return regex
+
+  @staticmethod
+  @ptransform_fn
+  def _matches_all_object(pcoll, regex, group=None):
+    """
+    PTransform to find the matches as per the regex.
+    """
+    def _process(element):
+      m = regex.finditer(element)
+      results = list()
+      for _, match in enumerate(m, start=1):
+        results.append(match.group())
+        for groupNum in range(0, len(match.groups())):
+          results.append(match.group(groupNum + 1))
+      if results:
+        yield results
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @ptransform_fn
+  def _matches_kv_object(pcoll, regex, keyGroup, valueGroup):
+    """
+    PTransfrom to find the matches as per regex and return as a KV pair.
+    """
+    def _process(element):
+      match = regex.match(element)
+      if match:
+        yield (match.group(keyGroup), match.group(valueGroup))
+
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def matches(pcoll, regex, group=None):
+    """
+    Returns the matches if the entire line matched the Regex. Returns the
+    entire line (group 0 by default). Group can be integer value and a string
+    value.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+      group: (optional) name of the group, it can be integer or a string value.
+    """
+    regex = Regex._regex_compile(regex)
+    group = group or 0
+
+    def _process(element):
+      m = regex.match(element)
+      if m:
+        yield m.group(group)
+    return pcoll | FlatMap(_process)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(str)
+  @ptransform_fn
+  def all_matches(pcoll, regex):
+    """
+    Returns the matches if the entire line matches the Regex. Returns all
+    groups.
+
+    Args:
+      regex: the regular expression string or (re.compile) pattern.
+    """
+    regex = Regex._regex_compile(regex)
+    return pcoll | Regex._matches_all_object(regex=regex)
+
+  @staticmethod
+  @typehints.with_input_types(str)
+  @typehints.with_output_types(typehints.KV[K, V])
+  @ptransform_fn
+  def matches_kv(pcoll, regex, keyGroup, valueGroup):
+    """
+    Returns the matches if the entire line matches the Regex. Returns the
```

Comment:
"Returns the matches" doesn't seem right. I'd say it maps the string into a kv pair if the string matches a regular expression, deriving the key and value from the specified group of the regular expression.

Reply:
Done